# MOwNiT laboratorium nr 1 - Arytmetyka komputerowa

Standard IEEE-754 http://babbage.cs.qc.cuny.edu/IEEE-754/

### Zadanie 1
Wykorzystaj funkcję finfo z biblioteki numpy aby określić precyzję i zakres różnych typów danych reprezentujących liczby zmienno-przecinkowe. Wyniki badania przedstaw w postaci zestawienia tabelrycznego. 

In [5]:
import numpy as np

print 'float16: min:', np.finfo(np.float16).min, ' max:', np.finfo(np.float16).max, ' precyzja:', np.finfo(np.float16).precision
print 'float32: min:', np.finfo(np.float32).min, ' max:', np.finfo(np.float32).max, ' precyzja:', np.finfo(np.float32).precision
print 'float64: min:', np.finfo(np.float64).min, ' max:', np.finfo(np.float64).max, ' precyzja:', np.finfo(np.float64).precision
print 'float128: min:', np.finfo(np.float128).min, ' max:', np.finfo(np.float128).max, ' precyzja:', np.finfo(np.float128).precision

float16: min: -65500.0  max: 65500.0  precyzja: 3
float32: min: -3.4028235e+38  max: 3.4028235e+38  precyzja: 6
float64: min: -1.7976931348623157e+308  max: 1.7976931348623157e+308  precyzja: 15
float128: min: -1.189731495357231765e+4932  max: 1.189731495357231765e+4932  precyzja: 18


Typ | Minimum | Maksimum | Precyzja
--- | --- | --- | ---
float16 | -3.4028235e+38  | 3.4028235e+38 | 3
float32 | -3.4028235e+38 | 3.4028235e+38 | 6
float64 | -1.7976931348623157e+308 | 1.7976931348623157e+308 | 15
float128 | -1.189731495357231765e+4932 | 1.189731495357231765e+4932 | 18

### Zadanie 2 
Napisz funkcje, która dla dowolnej liczby zmiennoprzecinkowej w precyzji 64-bitowej (w standardzie IEEE 754) przedstawi w wersji 
dziesiętnej i binarnej składniki liczby: znak, wykładnik i mantysa. Można wykorzystać między innymi bibliotekę bitstring. Zbadaj jak reprezentowaną są poszczególne składniki tego formatu: zakres, sposób kodowania liczby etc. 

In [15]:
import bitstring as bs

def toDec(x, p):
    res = 0
    for i in x[0:]:
        if i == '1':
            res += 2 ** p
        p -= 1
    return res

def num_info(number):
    bf = bs.BitArray(float=number, length=64)
    
    print 'Składniki liczby w postaci binarnej'
    print 'Znak:', bf.bin[0]
    print 'Cecha:', bf.bin[1:12]
    mantysa = '1.'+bf.bin[12:64]
    print 'Mantysa:', mantysa, '\n'
    
    print 'Składniki liczby w postaci dziesiętnej'
    print 'Znak: ', '+' if bf.bin[0]=='0' else '-'
    print 'Cecha:', int(bf.bin[1:12], 2)
    print 'Mantysa:', 1+toDec(bf.bin[12:64], -1)    
    
num_info(110.375)

Składniki liczby w postaci binarnej
Znak: 0
Cecha: 10000000101
Mantysa: 1.1011100110000000000000000000000000000000000000000000 

Składniki liczby w postaci dziesiętnej
Znak:  +
Cecha: 1029
Mantysa: 1.724609375


### Zadanie 3 
Wykorzystaj funkcję nextafter() z biblioteki numpy (lub inny sposób) i sprawdź rozdzielczość typu float (jakie mają wartości najbliższe sąsiadujące liczby float). Zaprezentuj wynik korzystając z funkcji z Zadania 2. 

In [19]:
import numpy as np

x = num_info(np.nextafter(0, 1))
print(x)

Składniki liczby w postaci binarnej
Znak: 0
Cecha: 00000000000
Mantysa: 1.0000000000000000000000000000000000000000000000000001 

Składniki liczby w postaci dziesiętnej
Znak:  +
Cecha: 0
Mantysa: 1.0
None


### Zadanie 5
Sprawdź czym jest postać znormalizowana liczby typu float. Poniższy program generuje ciąg liczb. Wyświetl ich postać binarną i sprawdź czy są w postaci znormalizowanej (napisz stosowny kod, który to sprawdza). 

Postać **_znormalizowana_** liczby typu float to taka postać, w której mantysa reprezentuje liczbę w postaci 1,xxx...xxx, a nie 0,xxx...xxxx

In [23]:
def decimalToBinary(num): 
  
    binary = ""    
    integral = int(num)  
    fractional = num - integral
    
    while (integral) :   
        rem = integral % 2 
        binary += str(rem);  
        integral //= 2
    
    if integral == 0:
        binary = '0'
    else:
        binary = binary[ : : -1]  
    
    binary += '.'
    
    while (fractional > 0): 
        fractional *= 2
        fract_bit = int(fractional)  
        
        if (fract_bit == 1):  
            fractional -= fract_bit  
            binary += '1'
        else : 
            binary += '0'
            
    return binary

def print_info(a, b):
    print a, b, 'znormalizowana' if b[0] == 0 and b[1] == '.' else 'nieznormalizowana'

a=np.float32(1.1)
for i in range(0, 149):
    a=a/np.float32(2.0)
    b = decimalToBinary(a)
    print_info(a,b)
    
    

 0.55 0.100011001100110011001101 nieznormalizowana
0.275 0.0100011001100110011001101 nieznormalizowana
0.1375 0.00100011001100110011001101 nieznormalizowana
0.06875 0.000100011001100110011001101 nieznormalizowana
0.034375 0.0000100011001100110011001101 nieznormalizowana
0.0171875 0.00000100011001100110011001101 nieznormalizowana
0.00859375 0.000000100011001100110011001101 nieznormalizowana
0.004296875 0.0000000100011001100110011001101 nieznormalizowana
0.0021484375 0.00000000100011001100110011001101 nieznormalizowana
0.0010742188 0.000000000100011001100110011001101 nieznormalizowana
0.0005371094 0.0000000000100011001100110011001101 nieznormalizowana
0.0002685547 0.00000000000100011001100110011001101 nieznormalizowana
0.00013427735 0.000000000000100011001100110011001101 nieznormalizowana
6.713867e-05 0.0000000000000100011001100110011001101 nieznormalizowana
3.3569337e-05 0.00000000000000100011001100110011001101 nieznormalizowana
1.6784668e-05 0.000000000000000100011001100110011001101 ni

### Zadanie 4
Porównać reprezentację bitową liczby 1/3 dla typów float o różnej precyzji: float16, float32, float64. Sprawdź co się dzieje gdy zmienna o mniejszej precyzji konwertowana jest do wyższej. 

In [32]:
import bitstring

f16 = (np.float16)(1.0/3.0)
print 'reprezentacja dziediętna:', f16, ', binarna:', decimalToBinary(f16)

f32 = (np.float32)(1.0/3.0)
print 'reprezentacja dziediętna:', f32, ', binarna:', decimalToBinary(f32)

f64 = (np.float64)(1.0/3.0)
print 'reprezentacja dziediętna:', f64, ', binarna:', decimalToBinary(f64)

f = (np.float64)(f32)
print 'reprezentacja binarna float32 skonwertowanego do float64:', '0.' + bitstring.BitArray(float=f, length=64).bin

reprezentacja dziediętna: 0.3333 , binarna: 0.010101010101
reprezentacja dziediętna: 0.33333334 , binarna: 0.0101010101010101010101011
reprezentacja dziediętna: 0.3333333333333333 , binarna: 0.010101010101010101010101010101010101010101010101010101
reprezentacja binarna float16 skonwertowanego do float64: 0.0011111111010101010101010101010101100000000000000000000000000000


Jak widzimy przy konwertowaniu float'a o mniejszej precyzji do wyższej dodatkowe bity, który pojawiły się przy zwiększeniu precyzji są zerami. Więc nie ma sensu konwertować z mniejszej precyzjido wyższej.

### Zadanie 6
Wyjaśnij dlaczego poniższy program nie działa prawidłowo. Sprawdź co się stanie gdy użyjemy typów o mniejszej precyzji. Zaproponuj inny sposób sumowania, który będzie dawał dokładne wyniki. 

Poniższy program nie działą prawidłowo, bo przy mnożeniu zaokrąglenie liczby jest wykonywane jeden raz, a przy wykonywaniu operacji dodawania w pętli N razy.

In [65]:
N = 100000000
C = 0.09531258654533566;

sum = (np.float128)(0.0)
for i in range(0, N):
    sum += C
    
print(sum)
print(N*C)

9531258.654533305327
9531258.65453


Dla sumowania liczb można stosować algorytm _Kahana_, który daje dokłądniejszy wynik. Algorytm Kahana jest przedstawiony poniżej

In [37]:
N = 100000000
C = 0.09531258654533566;

def kahansum(nbr, input):
    summ = c = 0
    for i in range(0, input):
        nbr_ = nbr - c
        sum_ = summ + nbr_
        c = (sum_ - summ) - nbr_
        summ = sum_
    return summ

print(kahansum(C, N))
print(N*C)



9531258.65453
9531258.65453
